File: sdklibrary/src/main/java/com/loner/android/sdk/webservice/network/apis/RegisterRequest.java
package com.loner.android.sdk.webservice.network.apis;

import com.loner.android.sdk.model.request.ActivationCode;
import com.loner.android.sdk.model.response.AccessToken;
import com.loner.android.sdk.utilities.Constant;
import com.loner.android.sdk.webservice.interfaces.HTTPClientInterface;
import com.loner.android.sdk.webservice.network.helper.NetworkConstants;
import com.loner.android.sdk.webservice.network.helper.NetworkErrorInformation;
import com.loner.android.sdk.webservice.network.helper.NetworkSuccessInformation;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class RegisterRequest extends BaseRequest {

    private static final String TAG = RegisterRequest.class.getSimpleName();

    private HTTPClientInterface listener;
    private Call<AccessToken> callRegistration;
    private ActivationCode activationCode;

    public RegisterRequest(String accessToken, HTTPClientInterface listener) {
        this.listener = listener;
        activationCode = new ActivationCode(accessToken);
    }

    public void register(final BaseRequest requestObject, final int STASK_Authentication) {
        callRegistration = interfaceAPI.register(activationCode);
        callRegistration.enqueue(new Callback<AccessToken>() {
            @Override
            public void onResponse(Call<AccessToken> call, Response<AccessToken> response) {
                if (response.code() == 200) {
                    AccessToken accessToken = response.body();
                    NetworkSuccessInformation networkSuccessInformation = new NetworkSuccessInformation(STASK_Authentication);
                    listener.onSuccess(requestObject, accessToken, networkSuccessInformation);
                } else {
                    NetworkErrorInformation errorInformation = new NetworkErrorInformation();
                    errorInformation.setDetailMessage(NetworkConstants.FAILURE_PAYLOAD);
                    listener.onFailure(requestObject, errorInformation);
                }

            }

            @Override
            public void onFailure(Call<AccessToken> call, Throwable t) {
                NetworkErrorInformation errorInformation = new NetworkErrorInformation();
                errorInformation.setDetailMessage(NetworkConstants.FAILURE_PAYLOAD);
                listener.onFailure(requestObject, errorInformation);
                callRegistration.cancel();
            }
        });
    }
}

File: sdklibrary/src/main/java/com/loner/android/sdk/webservice/network/apis/RegisterRequest.kt
package com.loner.android.sdk.webservice.network.apis

import com.loner.android.sdk.model.request.ActivationCode
import com.loner.android.sdk.model.response.AccessToken
import com.loner.android.sdk.webservice.interfaces.HTTPClientInterface
import com.loner.android.sdk.webservice.network.helper.NetworkConstants
import com.loner.android.sdk.webservice.network.helper.NetworkErrorInformation
import com.loner.android.sdk.webservice.network.helper.NetworkSuccessInformation
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class RegisterRequest(accessToken: String, private val listener: HTTPClientInterface) : BaseRequest() {
    private var callRegistration: Call<AccessToken>? = null
    private val activationCode: ActivationCode = ActivationCode(accessToken)

    fun register(requestObject: BaseRequest, STASK_Authentication: Int) {
        callRegistration = interfaceAPI.register(activationCode)
        callRegistration!!.enqueue(object : Callback<AccessToken> {
            override fun onResponse(call: Call<AccessToken>, response: Response<AccessToken>) {
                if (response.code() == 200) {
                    val accessToken = response.body()
                    val networkSuccessInformation = NetworkSuccessInformation(STASK_Authentication)
                    listener.onSuccess(requestObject, accessToken!!, networkSuccessInformation)
                } else {
                    val errorInformation = NetworkErrorInformation()
                    errorInformation.error = "Error"
                    errorInformation.detailMessage = NetworkConstants.FAILURE_PAYLOAD
                    listener.onFailure(requestObject, errorInformation)
                }

            }

            override fun onFailure(call: Call<AccessToken>, t: Throwable) {
                val errorInformation = NetworkErrorInformation()
                errorInformation.error = "Error"
                errorInformation.detailMessage = NetworkConstants.FAILURE_PAYLOAD
                listener.onFailure(requestObject, errorInformation)
                callRegistration!!.cancel()
            }
        })
    }

    companion object {
        private val TAG = RegisterRequest::class.java.simpleName
    }
}